### ISSUE TITLE
Getting wrong predictions after saving and reloading an Estimator

### ISSUE THREAD

**Issue by psmlbhor** [tensorflow/tensorflow#24857]
I am trying to save an `Estimator` and then load it to predict as required. Part where I train the model:

    classifier = tf.estimator.Estimator(model_fn=bag_of_words_model)
    
    # Train
    train_input_fn = tf.estimator.inputs.numpy_input_fn(
        x={"words": x_train},  # x_train is 2D numpy array of shape (26, 5)
        y=y_train,                   # y_train is 1D panda series of length 26
        batch_size=1000,
        num_epochs=None,
        shuffle=True)
    
    classifier.train(input_fn=train_input_fn, steps=300)

I then save the model as follows:

    def serving_input_receiver_fn():
        serialized_tf_example = tf.placeholder(dtype=tf.int64, shape=(None, 5), name='words')
        receiver_tensors = {"predictor_inputs": serialized_tf_example}
        features = {"words": tf.tile(serialized_tf_example, multiples=[1, 1])}
        return tf.estimator.export.ServingInputReceiver(features, receiver_tensors)
    
    full_model_dir = classifier.export_savedmodel(export_dir_base="E:/models/",
                                                  serving_input_receiver_fn=serving_input_receiver_fn)

I now load the model and give the test set to it for prediction:

    from tensorflow.contrib import predictor
    
    classifier = predictor.from_saved_model("E:\\models\\1547122667")
    predictions = classifier({'predictor_inputs': x_test})
    print(predictions)

This gives me predictions like:

    {'class': array([ 0,  0,  0,  0,  0,  5,  0,  0,  0,  0,  0,  0,  0,  0,  0, 15,  0,
            0,  5,  0, 20,  0,  5,  0,  0,  0], dtype=int64),
    'prob': array([[9.9397606e-01, 6.5355714e-05, 2.2225287e-05, ..., 1.4510043e-07,
                1.6920333e-07, 1.4865007e-07],
               [9.9886864e-01, 1.4976941e-06, 7.0847680e-05, ..., 9.4182191e-08,
                1.1828639e-07, 9.5683227e-08],
               [9.9884748e-01, 2.1105163e-06, 1.1994909e-05, ..., 8.3957858e-08,
                1.0476184e-07, 8.5592234e-08],
               ...,
               [9.6145850e-01, 6.9048328e-05, 1.1446012e-04, ..., 7.3761731e-07,
                8.8173107e-07, 7.3824998e-07],
               [9.7115618e-01, 2.9716679e-05, 5.9592247e-05, ..., 2.8933655e-07,
                3.4183532e-07, 2.9737942e-07],
               [9.7387028e-01, 6.9163914e-05, 1.5800977e-04, ..., 1.6116818e-06,
                1.9025001e-06, 1.5990496e-06]], dtype=float32)}

`class` and `prob` are two things that I am predicting. Now, if I predict the output with the same test set without saving and loading the model:

    # Predict.
    test_input_fn = tf.estimator.inputs.numpy_input_fn(
        x={"words": x_test}, y=y_test, num_epochs=1, shuffle=False)
    predictions = classifier.predict(input_fn=test_input_fn)
    print(predictions)

then I get the output as follows:

    {'class': 0, 'prob': array([9.9023646e-01, 2.6038184e-05, 3.9950578e-06, ..., 1.3950405e-08,
           1.5713249e-08, 1.3064114e-08], dtype=float32)}
    {'class': 1, 'prob': array([2.0078469e-05, 9.9907070e-01, 8.9245419e-05, ..., 6.6533559e-08,
           7.1365662e-08, 6.8764685e-08], dtype=float32)}
    {'class': 2, 'prob': array([3.0828053e-06, 9.6484597e-05, 9.9906868e-01, ..., 5.9190391e-08,
           6.0995028e-08, 6.2322023e-08], dtype=float32)}
    {'class': 3, 'prob': array([7.4923842e-06, 1.1112734e-06, 1.1697492e-06, ..., 4.4295877e-08,
           4.4563325e-08, 4.0475427e-08], dtype=float32)}
    {'class': 4, 'prob': array([4.6085161e-03, 2.8403942e-05, 2.0638861e-05, ..., 7.6083229e-09,
           8.5255349e-09, 6.7836012e-09], dtype=float32)}
    {'class': 5, 'prob': array([6.2119620e-06, 7.2357750e-07, 2.6231232e-06, ..., 7.4999367e-09,
           9.0847436e-09, 7.5630142e-09], dtype=float32)}
    {'class': 6, 'prob': array([4.4882968e-06, 2.2007227e-06, 8.3352124e-06, ..., 2.3130213e-09,
           2.3657243e-09, 2.0045692e-09], dtype=float32)}
    {'class': 7, 'prob': array([1.88617545e-04, 9.01482690e-06, 1.47353385e-05, ...,
           3.38567552e-09, 3.97709154e-09, 3.37017392e-09], dtype=float32)}
    {'class': 8, 'prob': array([1.9843496e-06, 4.5909755e-06, 4.8804057e-05, ..., 2.2636470e-08,
           2.0094852e-08, 2.0215294e-08], dtype=float32)}
    {'class': 9, 'prob': array([2.5907659e-04, 4.4661370e-05, 6.9490757e-06, ..., 1.6249915e-08,
           1.7579131e-08, 1.5439820e-08], dtype=float32)}
    {'class': 10, 'prob': array([3.6456138e-05, 7.5861579e-05, 3.0208937e-05, ..., 2.7859956e-08,
           2.5423596e-08, 2.8662368e-08], dtype=float32)}
    {'class': 11, 'prob': array([1.1723863e-05, 9.1407037e-06, 4.8835855e-04, ..., 2.3693143e-08,
           2.0524153e-08, 2.3223269e-08], dtype=float32)}
    {'class': 12, 'prob': array([1.2886175e-06, 2.6652628e-05, 2.7812246e-06, ..., 4.8295210e-08,
           4.4282604e-08, 4.7342766e-08], dtype=float32)}
    {'class': 13, 'prob': array([3.3486103e-05, 1.3361238e-05, 3.6493871e-05, ..., 2.2195401e-09,
           2.4768412e-09, 2.0150714e-09], dtype=float32)}
    {'class': 14, 'prob': array([4.6108948e-05, 3.0377207e-05, 2.0945006e-06, ..., 4.2276231e-08,
           5.2376720e-08, 4.4969173e-08], dtype=float32)}
    {'class': 15, 'prob': array([1.7165689e-04, 2.9350400e-05, 3.2283624e-05, ..., 7.1849078e-09,
           7.6871531e-09, 6.6224697e-09], dtype=float32)}
    {'class': 16, 'prob': array([5.9876328e-07, 3.0931276e-06, 1.5760432e-05, ..., 4.0450086e-08,
           4.2720632e-08, 4.6017195e-08], dtype=float32)}
    {'class': 17, 'prob': array([2.6658317e-04, 9.9656281e-05, 4.0355867e-06, ..., 1.2873563e-08,
           1.4808875e-08, 1.2155732e-08], dtype=float32)}
    {'class': 18, 'prob': array([1.4914459e-04, 2.1025437e-06, 1.2505146e-05, ..., 9.8899635e-09,
           1.1115599e-08, 8.9312255e-09], dtype=float32)}
    {'class': 19, 'prob': array([2.5615416e-04, 2.3750392e-05, 2.2886352e-04, ..., 3.9635733e-08,
           4.5139984e-08, 3.8605780e-08], dtype=float32)}
    {'class': 20, 'prob': array([6.3949975e-04, 2.3652929e-05, 7.8577641e-06, ..., 2.0959168e-09,
           2.5495863e-09, 2.0428985e-09], dtype=float32)}
    {'class': 21, 'prob': array([8.2179489e-05, 8.4409467e-06, 5.4756888e-06, ..., 2.2360982e-09,
           2.4820561e-09, 2.1206517e-09], dtype=float32)}
    {'class': 22, 'prob': array([3.9681905e-05, 2.4394642e-06, 8.9102805e-06, ..., 2.0282410e-08,
           2.1132811e-08, 1.8368105e-08], dtype=float32)}
    {'class': 23, 'prob': array([3.0794261e-05, 6.5104805e-06, 3.3528936e-06, ..., 2.0360846e-09,
           1.9360573e-09, 1.7195430e-09], dtype=float32)}
    {'class': 24, 'prob': array([3.4596618e-05, 2.2907707e-06, 2.5318438e-06, ..., 1.1038886e-08,
           1.2148775e-08, 9.9556408e-09], dtype=float32)}
    {'class': 25, 'prob': array([1.4846727e-03, 1.9189476e-06, 5.3232620e-06, ..., 3.1966723e-09,
           3.5612517e-09, 3.0947123e-09], dtype=float32)}

which is correct. Notice the difference between two outputs is that the `class` in the second one is increasing 1 by 1 while the `class` in the first case shows 0s at most places.

Why is there a difference in the prediction? Am I saving the model in a wrong way?

@ispirmustafa @random-forests @DynamicWebPaige

**Comment by psmlbhor** [comment]
I have also asked a question on stackover https://stackoverflow.com/questions/54129098/getting-wrong-prediction-after-loading-a-saved-model

**Comment by psmlbhor** [comment]
I tried it but I got same results.

```
from tensorflow.contrib.estimator import SavedModelEstimator
classifier = SavedModelEstimator("E:\\models\\1547539334")
test_input_fn = tf.estimator.inputs.numpy_input_fn(
    x={WORDS_FEATURE: x_test}, y=y_test, num_epochs=1, shuffle=False)
predictions = classifier.predict(input_fn=test_input_fn)
for p in predictions:
    print(p)
```

This gives:

```
{'class': 0, 'prob': array([9.9717832e-01, 8.5786020e-04, 6.4049636e-05, ..., 8.6544226e-08,
       8.6821146e-08, 8.1394326e-08], dtype=float32)}
{'class': 0, 'prob': array([9.9960703e-01, 2.0104484e-05, 1.6036544e-04, ..., 3.1367048e-08,
       3.2443833e-08, 3.0766806e-08], dtype=float32)}
{'class': 0, 'prob': array([9.9885404e-01, 7.5795266e-05, 6.7050540e-05, ..., 1.3488129e-07,
       1.4307327e-07, 1.3248858e-07], dtype=float32)}
{'class': 0, 'prob': array([9.9899799e-01, 6.4979497e-05, 4.9802653e-05, ..., 4.6777117e-08,
       4.7913996e-08, 4.3893024e-08], dtype=float32)}
{'class': 0, 'prob': array([9.8838729e-01, 3.7744928e-03, 2.5661272e-04, ..., 7.3713346e-07,
       7.7911784e-07, 7.5461628e-07], dtype=float32)}
{'class': 5, 'prob': array([1.2085142e-05, 1.5967547e-05, 1.6229365e-05, ..., 1.8335292e-07,
       2.0694857e-07, 2.1337026e-07], dtype=float32)}
{'class': 0, 'prob': array([9.8932624e-01, 3.0042315e-04, 3.4193668e-04, ..., 1.1369635e-06,
       1.2429477e-06, 1.1229140e-06], dtype=float32)}
{'class': 0, 'prob': array([9.8753148e-01, 1.1704578e-03, 4.1641024e-04, ..., 4.1676759e-07,
       4.4668820e-07, 3.9701357e-07], dtype=float32)}
{'class': 0, 'prob': array([9.8414856e-01, 4.8066137e-04, 4.8338709e-04, ..., 8.6252527e-07,
       8.3684807e-07, 7.9297627e-07], dtype=float32)}
{'class': 0, 'prob': array([9.8127502e-01, 4.0246081e-04, 9.1002218e-04, ..., 9.3697497e-07,
       9.7139502e-07, 9.0600975e-07], dtype=float32)}
{'class': 0, 'prob': array([9.9923956e-01, 4.5225835e-05, 3.3328299e-05, ..., 6.3075376e-08,
       6.8231451e-08, 6.2429535e-08], dtype=float32)}
{'class': 0, 'prob': array([9.9892765e-01, 6.2995081e-05, 8.0713442e-05, ..., 8.5985938e-08,
       9.5674274e-08, 8.6504272e-08], dtype=float32)}
{'class': 0, 'prob': array([9.9878377e-01, 1.2373565e-04, 5.3474701e-05, ..., 1.2326316e-07,
       1.2942408e-07, 1.1582518e-07], dtype=float32)}
{'class': 0, 'prob': array([9.8602110e-01, 5.0440140e-04, 3.7236928e-04, ..., 1.1675936e-06,
       1.2900981e-06, 1.1921265e-06], dtype=float32)}
{'class': 0, 'prob': array([9.99113023e-01, 1.20989556e-04, 4.73294276e-05, ...,
       6.17745854e-08, 6.85775419e-08, 6.50879883e-08], dtype=float32)}
{'class': 15, 'prob': array([7.3899527e-04, 7.2398130e-03, 1.9108003e-03, ..., 1.9992517e-07,
       2.1110104e-07, 2.0947941e-07], dtype=float32)}
{'class': 0, 'prob': array([9.9912304e-01, 4.3242489e-05, 3.3706045e-05, ..., 5.1705765e-08,
       5.6073219e-08, 4.9725138e-08], dtype=float32)}
{'class': 0, 'prob': array([9.7001451e-01, 5.3095614e-04, 2.6374252e-03, ..., 2.1944559e-06,
       2.3350731e-06, 2.0890418e-06], dtype=float32)}
{'class': 18, 'prob': array([7.0199365e-04, 6.0002081e-04, 7.1425311e-04, ..., 3.8498611e-06,
       4.1461471e-06, 3.8720273e-06], dtype=float32)}
{'class': 0, 'prob': array([9.9898738e-01, 4.7222173e-05, 7.6803917e-05, ..., 8.8532246e-08,
       9.0705051e-08, 8.4998817e-08], dtype=float32)}
{'class': 18, 'prob': array([6.7510635e-02, 4.0915180e-03, 6.6295490e-03, ..., 1.0756982e-05,
       1.0628254e-05, 1.0272101e-05], dtype=float32)}
{'class': 0, 'prob': array([9.7325921e-01, 9.5799292e-04, 9.8795001e-04, ..., 1.9478718e-07,
       2.0265945e-07, 1.8819739e-07], dtype=float32)}
{'class': 5, 'prob': array([5.7400834e-02, 7.8264829e-03, 9.8085916e-03, ..., 3.3320775e-05,
       3.6073146e-05, 3.5094916e-05], dtype=float32)}
{'class': 0, 'prob': array([9.7876966e-01, 1.3300353e-03, 1.1287389e-03, ..., 6.9847374e-07,
       7.3241597e-07, 6.7505181e-07], dtype=float32)}
{'class': 0, 'prob': array([9.7674805e-01, 6.3739752e-04, 1.0363439e-03, ..., 1.0288414e-06,
       1.0741144e-06, 1.0069001e-06], dtype=float32)}
{'class': 0, 'prob': array([9.8600847e-01, 6.5606064e-04, 6.1158347e-04, ..., 1.3908433e-06,
       1.4873067e-06, 1.3897480e-06], dtype=float32)}
```

Most of the classes are again ```0```.

**Comment by karmel** [comment]
Are either of these known correct? My suspicion here is that Estimator.predict is giving you the wrong values, and that both predictor and SavedModelEstimator are correct, as the class values from .predict seem to be just incrementing. Can you provide the Estimator code in question? @bananabowl , can you take a look?

**Comment by karmel** [comment]
(To clarify, I meant provide the model_fn, bag_of_words_model)

**Comment by psmlbhor** [comment]
Sure. Here you go:

```
def bag_of_words_model(features, labels, mode):
    """A bag-of-words model. Note it disregards the word order in the text."""
    bow_column = tf.feature_column.categorical_column_with_identity(
        WORDS_FEATURE, num_buckets=n_words)
    bow_embedding_column = tf.feature_column.embedding_column(
        bow_column, dimension=EMBEDDING_SIZE)
    bow = tf.feature_column.input_layer(
        features, feature_columns=[bow_embedding_column])
    logits = tf.layers.dense(bow, MAX_LABEL, activation=None)

    return estimator_spec_for_softmax_classification(
        logits=logits, labels=labels, mode=mode)

def estimator_spec_for_softmax_classification(logits, labels, mode):
  """Returns EstimatorSpec instance for softmax classification."""
  predicted_classes = tf.argmax(logits, 1)
  if mode == tf.estimator.ModeKeys.PREDICT:
    export_outputs = {'predict_output': tf.estimator.export.PredictOutput({"class": predicted_classes, 'prob': tf.nn.softmax(logits)})}
    return tf.estimator.EstimatorSpec(
        mode=mode,
        predictions={
            'class': predicted_classes,
            'prob': tf.nn.softmax(logits)
        },
        export_outputs=export_outputs)

  loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
  if mode == tf.estimator.ModeKeys.TRAIN:
    optimizer = tf.train.AdamOptimizer(learning_rate=0.01)
    train_op = optimizer.minimize(loss, global_step=tf.train.get_global_step())
    return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)

  eval_metric_ops = {
      'accuracy':
          tf.metrics.accuracy(labels=labels, predictions=predicted_classes)
  }
  return tf.estimator.EstimatorSpec(
      mode=mode, loss=loss, eval_metric_ops=eval_metric_ops)

```

> 
> 
> Are either of these known correct? My suspicion here is that Estimator.predict is giving you the wrong values, and that both predictor and SavedModelEstimator are correct, as the class values from .predict seem to be just incrementing.
Actually, my test set is exactly like the train set. Each entry belongs to different class.

**Comment by bananabowl** [comment]
Thanks for reporting this, I will take a look. Are you running TF 1.12.0?

**Comment by bananabowl** [comment]
Two theories:

1. The exported saved model is incorrect. You might be able to rule this out with the saved model CLI tool which you can use to manually determine the output of a given input to the model: https://www.tensorflow.org/guide/saved_model#cli_to_inspect_and_execute_savedmodel
If it produces the wrong output, we'll know there's some bug in the export step.

2. The x_test is somehow different between the two calls.

You may want to remove the y kwarg passed to numpy_input_fn when building test_input_fn as it is not needed in prediction. This shouldn't affect the output though.

**Comment by psmlbhor** [comment]
> 
>     2. The x_test is somehow different between the two calls.

Yes, this was indeed the case. I checked the model using `saved_model_cli` but got the same outputs as like before. I wasn't sure about `x_test` being wrong between 2 different calls but at last it seemed to be the only option. I found out that `x_test` which I was passing after loading the model had values +1 than required. I subtracted 1 from each value and it worked! Thank you very much @bananabowl  @karmel .
